External/FreeSASA/RDFreeSASA.cpp

```diff
@@ -0,0 +1,209 @@
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/QueryOps.h>
+#include <GraphMol/RDKitQueries.h>
+#include <GraphMol/MonomerInfo.h>
+#include <RDGeneral/types.h>
+#include "RDFreeSASA.h"
+#include "boost/format.hpp"
+
+extern "C" {
+#include "freesasa.h"
+}
+
+namespace RDKit {
+  namespace common_properties {
+    namespace Atom {
+      const std::string SASA = "SASA";; // Solvent Accessible Surface Area for atom- double
+      const std::string SASAClass = "SASAClass";     // Class type, 0,1,2... etc
+      const std::string SASAClassName = "SASAClassName"; // Class name, Polar, APolar etc...
+    }
+    namespace Molecule {
+      const std::string SASA = "SASA"; // Total Solvent Accessible Surface area for molecule;
+    }
+  }
+}
+
+namespace FreeSASA {
+using namespace RDKit;
+
+bool classifyAtoms(ROMol &mol, std::vector<double> &radii,
+                   const SASAOpts &opts) {
+  radii.clear();
+  const freesasa_classifier *classifier = 0;
+  switch (opts.d_classifier) {
+    case SASAOpts::Protor:
+      classifier = &freesasa_protor_classifier;
+      break;
+    case SASAOpts::NACCESS:
+      classifier = &freesasa_naccess_classifier;
+      break;
+    case SASAOpts::OONS:
+      classifier = &freesasa_oons_classifier;
+      break;
+    default:
+      throw ValueErrorException("unknown FreeSASA classifier specified");
+      return false;
+  }
+
+  bool success = true;
+  for (ROMol::AtomIterator at = mol.beginAtoms(); at != mol.endAtoms(); ++at) {
+    Atom *atom = *at;
+    int cls = -1;
+    std::string classification = "Unclassified";
+    double radius = 0.0;
+
+    const AtomMonomerInfo *info = atom->getMonomerInfo();
+    if (info) {
+      const char *atom_name = info->getName().c_str();
+      const char *res_name = 0;
+
+      if (info->getMonomerType() == AtomMonomerInfo::PDBRESIDUE) {
+        res_name = ((AtomPDBResidueInfo *)info)->getResidueName().c_str();
+        radius = classifier->radius(res_name, atom_name, classifier);
+
+        if (radius == 0.0) {
+          BOOST_LOG(rdWarningLog) << "Atom " << atom->getIdx()
+                                  << " has zero radius" << std::endl;
+        }
+
+        cls = classifier->sasa_class(res_name, atom_name, classifier);
+        if (cls == FREESASA_FAIL) {
+          std::string err = boost::str(
+              boost::format("Atom %d% failed classification") % atom->getIdx());
+          throw ValueErrorException(err);
+        } else if (cls == FREESASA_WARN) {
+          BOOST_LOG(rdWarningLog) << "Atom " << atom->getIdx()
+                                  << " could not be classified" << std::endl;
+          success = false;
+        } else {
+          classification = classifier->class2str(cls, classifier);
+        }
+      }
+    }
+
+    radii.push_back(radius);
```

Comment:
I think so.  The function returns true/false on failure.  I don't think that there is any harm in using 0 radii for unclassified atoms.